### PR TITLE
Correct GDASApp paths

### LINKED ISSUE #2434: incorrect GDASApp paths
### What is wrong?

GDASApp issue [#991](https://github.com/NOAA-EMC/GDASApp/issues/991) documents incorrect paths in GDASApp g-w jobs.  The incorrect paths fall into two categories:
1. change in GDASApp directory structure not reflected in g-w job.  For example, `jobs/rocoto/prepsnowobs.sh` contains `${HOMEgfs}/sorc/gdas.cd/iodaconv/src`.   The correct path to `iodaconv` is now `gdas.cd/sorc/iodaconv`.
2. use different python versions on different machines.  For example, `jobs/rocoto/prepsnowobs.sh` references `python3.7` but the Hera Rocky 8 build uses `python3.10`.

### What should have happened?

g-w jobs should use the correct path to GDASApp components and python.

### What machines are impacted?

Hera, Orion, Hercules

### Steps to reproduce

1. Clone g-w `develop`. 
2. Compare GDASApp directory structure with assumed paths in `jobs/rocoto`
3. Build GDASApp
4. Compare GDASApp path to python in `build/lib` with assumed path in `jobs/rocoto`

### Additional information

Rahul identified this problem.

### Do you have a proposed solution?

Need to go through g-w `develop` and identify all the places where GDASApp paths are referenced.  Ensure paths are correct and GDASApp jobs run on Hera, Orion, and Hercules.   These three machine are, at present, the only machines on which GDASApp can be built and run.  

WCOSS2 Acorn has a build capability but this is a test system.  GDASApp can not yet be built on Cactus or Dogwood due to required libraries not being available.

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
# Description
The changes in this PR 
- account for changes in GDASApp directory structure
- generalize how the path to the GDASApp python ioda library is specified

Resolves #2434

# Type of change
- Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? **YES**, this change resolves problems with JEDI observation processing jobs
- Does this change require a documentation update? **NO**

# How has this been tested?
- [x] Clone and build on Hera and Orion
- [x] Run JEDI pieces of CI case `C96C48_ufs_hybatmDA.yaml` on Hera and Orion 

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP]: Instead of repeating ourselves, let's do this:

Add a new function to `ush/preamble.sh`:
```
detect_py_ver() {
    # 
    # Returns the major.minor version of the currently active python executable
    #
    regex="[0-9]+\.[0-9]+"
    if [[ $(python --version) =~ ${regex} ]]; then
	    echo "${BASH_REMATCH[0]}"
    else
	    echo "FATAL ERROR: Could not detect the python version"
	    exit 1
    fi
}
```

Then these all become:
```suggestion
pyiodaPATH="${HOMEgfs}/sorc/gdas.cd/build/lib/python$(detect_py_ver)/"
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
Comment (WalterKolczynski-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP]: We're going to be moving the growing number of bash functions to a new location soon, but for now preamble is the correct place.

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
Comment (RussTreadon-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP]: Add function `detect_py_ver` to `ush/preamble.sh`.  Invoke function to set `py_ver`.  Add `#shellcheck disable=SC2312` before line in question.  Done at c365e992

File: jobs/rocoto/prepatmiodaobs.sh
Comment (RussTreadon-NOAA) [reply jobs/rocoto/prepatmiodaobs.sh]: Add function `detect_py_ver` to `ush/preamble.sh`.  Invoke function to set `py_ver`.  Add `#shellcheck disable=SC2312` before line in question.  Done at c365e992

File: jobs/rocoto/prepsnowobs.sh
Comment (RussTreadon-NOAA) [reply jobs/rocoto/prepsnowobs.sh]: Add function `detect_py_ver` to `ush/preamble.sh`.  Invoke function to set `py_ver`.  Add `#shellcheck disable=SC2312` before line in question.  Done at c365e992

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
Comment (RussTreadon-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP]: Thank you very much @WalterKolczynski-NOAA for your guidance.  This is an excellent suggestion - much cleaner, compact, and centralized.  Suggested refactor committed to `RussTreadon-NOAA:feature/gdasapp_paths` at c365e992

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
Comment (RussTreadon-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP:38]: Add `# shellcheck disable=SC2311` at a9fd4469.  Not sure if this is the best way to address this.

File: jobs/rocoto/prepatmiodaobs.sh
Comment (RussTreadon-NOAA) [reply jobs/rocoto/prepatmiodaobs.sh:18]: Add `# shellcheck disable=SC2311` at a9fd4469.  Not sure if this is the best way to address this.

File: jobs/rocoto/prepsnowobs.sh
Comment (RussTreadon-NOAA) [reply jobs/rocoto/prepsnowobs.sh:18]: Add `# shellcheck disable=SC2311` at a9fd4469.  Not sure if this is the best way to address this.

**Comment by RussTreadon-NOAA** [comment]
@jiaruidong2017 and @AndrewEichmann-NOAA :  this PR alters `jobs/rocoto/prepsnowobs.sh` and `jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP`.   Would you mind checking the changes and confirm that they look OK to you?  Thank you.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Comment by TerrenceMcGuinness-NOAA** [comment]
gfs build failed on Orion:  [build_gdas.log](https://gist.github.com/emcbot/a26be8335f43c7da16974022c18e58d5)

```icpc: error #10106: Fatal error in /apps/intel-2022.1.2/intel-2022.1.2/compiler/2022.0.2/linux/bin/intel64/../../bin/intel64/mcpcom, terminated by kill signal
compilation aborted for /work2/noaa/stmp/CI/ORION/2435/gfs/sorc/gdas.cd/bundle/fv3-jedi/src/mains/fv3jediControlPert.cc (code 1)
make[2]: *** [fv3-jedi/src/mains/CMakeFiles/fv3jedi_controlpert.x.dir/fv3jediControlPert.cc.o] Error 1
make[2]: Leaving directory `/work2/noaa/stmp/CI/ORION/2435/gfs/sorc/gdas.cd/build'
make[1]: *** [fv3-jedi/src/mains/CMakeFiles/fv3jedi_controlpert.x.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by RussTreadon-NOAA** [comment]
@TerrenceMcGuinness-NOAA :  Interesting. I successfully built GDASApp on Orion this morning.  The log file is in `build_gdas.log` in /work2/noaa/da/rtreadon/git/global-workflow/gdasapp_paths/sorc/logs.   The build is executed by executing `sorc/build_all.sh -u`

**Comment by RussTreadon-NOAA** [comment]
@WalterKolczynski-NOAA and @TerrenceMcGuinness-NOAA :  what can or should I do to help move this PR forward?   Is it possible the CI compilation ran out of memory or stack space?   `uname -a` for my Orion environment returns
```
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 767186
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) 131072
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 16384
cpu time               (seconds, -t) unlimited
max user processes              (-u) 4096
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```
I don't see any special settings in my `.bashrrc`.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Not sure, we just had a successful build from the develop branch in this CI pipeline just an hour ago too.  Since it simply unconfounded lets try build it one more time. I'll add the `ulimit -a` line in the build script as well.  Doing that now.

**Comment by RussTreadon-NOAA** [comment]
Thank you @TerrenceMcGuinness-NOAA .  Fingers crossed this build works.

**Comment by WalterKolczynski-NOAA** [comment]
Manual build worked for me, so hopefully CI makes it this time.

I will note that there are a TON of cmake warnings that make it difficult to tell whether things are proceeding correctly. And I'm sure NCO will yell about them.

**Comment by RussTreadon-NOAA** [comment]
Thank you @WalterKolczynski-NOAA for confirming that a manual build works.  

Implementing JEDI in operations presents many challenges.  This first is getting all the needed support libraries installed on WCOSS.  Compiler warnings are a serious concern.  Tagging @CoryMartin-NOAA , though he is already aware of this.

**Comment by TerrenceMcGuinness-NOAA** [comment]
```
+ ./build_all.sh -wgu
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
Building gsi_enkf, gdas, ww3prepost, ufs_utils, ufs, gsi_utils, gsi_monitor, gfs_utils, upp
Starting build_gsi_enkf.sh
Starting build_gdas.sh
Starting build_ww3prepost.sh
Starting build_ufs_utils.sh
Starting build_ufs.sh
Starting build_gsi_utils.sh
Starting build_gsi_monitor.sh
Starting build_gfs_utils.sh
Starting build_upp.sh
build_gsi_enkf.sh completed successfully!
build_ww3prepost.sh completed successfully!
build_ufs_utils.sh completed successfully!
build_gfs_utils.sh completed successfully!
build_gsi_monitor.sh completed successfully!
build_gsi_utils.sh completed successfully!
build_upp.sh completed successfully!
build_ufs.sh completed successfully!
build_gdas.sh completed successfully!
.... Build system finished ....
```
Build ran though in CI without issues this time.
CI is now running all the PR cases.
This wasn't reported here because sometimes Jenkins can not reach GitHub and the reporting out gets skipped.
You can see the progress below in the GitHub Merge section.
Manually updating the label to Running.

**Comment by TerrenceMcGuinness-NOAA** [comment]
@RussTreadon-NOAA CI couldn't update the label but we can see below in the "Merge" box that the **Run Tests** _Stage built successfully._  I'll update the label manually to PASS.

**Comment by RussTreadon-NOAA** [comment]
Thank you @TerrenceMcGuinness-NOAA .  Sorry that this is requiring so much manual intervention from you.  I appreciate your diligence in monitoring the progress.

**Comment by WalterKolczynski-NOAA** [comment]
Thanks @TerrenceMcGuinness-NOAA for monitoring this.

@RussTreadon-NOAA Do you want me to hold this for Jiarui or Andrew to approve? Otherwise I will merge ASAP as I think this is blocking other PRs from testing.

**Comment by RussTreadon-NOAA** [comment]
Thank you for asking @WalterKolczynski-NOAA .  I vote for moving ahead with the merge.

**Comment by jiaruidong2017** [comment]
I confirm it is good for the changes in jobs/rocoto/prepsnowobs.sh.

**Pull request merged**
